buffer/src/main/java/io/netty/buffer/api/internal/LeakDetection.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright 2021 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.buffer.api.internal;
+
+import io.netty.buffer.api.LeakInfo;
+import io.netty.buffer.api.LoggingLeakCallback;
+import io.netty.buffer.api.MemoryManager;
+import io.netty.util.SafeCloseable;
+import io.netty.util.internal.SystemPropertyUtil;
+import io.netty.util.internal.UnstableApi;
+
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.VarHandle;
+import java.util.Collection;
+import java.util.IdentityHashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Consumer;
+import java.util.stream.Stream;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Utility class for the leak detection parts that are static and shared system-wide.
+ */
+@UnstableApi
+public final class LeakDetection {
+    static volatile int leakDetectionEnabled;
```

Comment:
as it stands today this is only accessed in a `synchronized` block, which means we could just not use `volatile`